pdns/recursordist/taskqueue.cc

```diff
@@ -0,0 +1,122 @@
+/*
+ * This file is part of PowerDNS or dnsdist.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "taskqueue.hh"
+
+#include "logger.hh"
+#include "syncres.hh"
+
+extern std::shared_ptr<SyncRes::domainmap_t> g_initialDomainMap;
+
+namespace pdns
+{
+
+bool TaskQueue::empty() const
+{
+  return d_queue.empty();
+}
+
+size_t TaskQueue::size() const
+{
+  return d_queue.size();
+}
+
+void TaskQueue::push(const ResolveTask&& task)
+{
+  // Insertion fails if it's already there, no problem since we're already scheduled
+  // and the deadline would remain the same anyway.
+  auto result = d_queue.insert(std::move(task));
```

Comment:
I don't think that's true, the deadline is a fraction of the original TTL that doesn't change plus the time of the insertion, which will.

Reply:
This is an error in the calls I think. The deadline was supposed to be the original ttd, with the idea that it does not make a lot of sense to execute the scheduled task after that. But I see the calls use the "almost expired" point in time.